pkgs/tools/misc/android-tools/default.nix

```diff
@@ -9,11 +9,11 @@ in
 
 stdenv.mkDerivation rec {
   pname = "android-tools";
-  version = "31.0.3p1";
+  version = "33.0.3";
 
   src = fetchurl {
     url = "https://github.com/nmeum/android-tools/releases/download/${version}/android-tools-${version}.tar.xz";
-    sha256 = "1f2svy381r798hjinrc2xiwz13gkkqxfill343zvv8jqkn8rzxhf";
+    sha256 = "sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=";
```

Comment:
Whats the advantage over sha256?

Reply:
I don't think there's an advantage?

From https://ryantm.github.io/nixpkgs/builders/fetchers/:

> The hash is typically `sha256`, although many more hash algorithms are supported. Nixpkgs contributors are currently recommended to use `sha256`.

Reply:
Okay sounds good, stayed with sha256 in my PR (#199212)

Reply:
I should've been more clear. `sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=` is a SRI hash. The hash itself includes the algorithm that is used for the hash.
These kinds of hashes are meant for `hash =` attributes.
`sha256 = ` should include a sha256 hash (one without the `sha256-` prefix). Apparently a SRI hash is also allowed, but it seems strange to use it for `sha256 = ` attributes.

See https://nixos.wiki/wiki/Nix_Hash#Usage, where the examples shows a SRI hash for the `hash` attribute and a sha256 hash for the `sha256` attribute.

So my suggestion is to use:

```
hash = "sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=";
```

or

```
sha256 = "jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=";
```
One isn't better than the other, but mixing the two seemed a bit strange.

Sorry I wasn't clear before and to make a 'discussion' out of a minor thing.

Reply:
No Problem @bobvanderlinden I'm open to learn/improve as I'm not this deep into it :)

As my current workflow mostly is using lib.fakeSha256 which will output the hashes like `sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=` it make sense to avoid the doubling of the algorithm and use `hash` instead of `sha256`. Which does not really interfere with the quote from @SuperSamus because it's focused on the used hash algorithm not the option specifying it. Did I get this correct?

Reply:
Haha, thanks for that attitude. To further the learning:

`lib.fakeSha256` outputs `0000000000000000000000000000000000000000000000000000000000000000`. `lib.fakeHash` outputs `sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=`.

That's why it would be strange to me to see `hash = lib.fakeSha256` or `sha256 = lib.fakeHash`. It should be `sha256 = lib.fakeSha256` or hash = lib.fakeHash`. It's interesting that `sha256 = lib.fakeHash` works at all; I'm almost inclined to say that this usage should be technically disallowed (assertion error).

See the implementations of the `fakeHash` and `fakeSha256`:
https://cs.github.com/NixOS/nixpkgs/blob/a2a777538d971c6b01c6e54af89ddd6567c055e8/lib/deprecated.nix#L304

Reply:
I'm kind of more confused now because I see you are right about the fakeSha256 and fakeHash-values but:

If I use fakeSha256 in android-tools e.g. my error-message (I'm using colmena if it is relevant) looks like:
```
error: hash mismatch in fixed-output derivation '/nix/store/dpb2xsnr0ya8qym3rxf6z5n2fykcygzl-android-tools-33.0.3.tar.xz.drv':
         specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
            got:    sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=
```

So that's why I came up using this value/format. To mix it up like using sha256 = lib.fakeHash sound really wrong and I could see a benefits for newer users to technically disallowing it.

Reply:
Okay formats in error-message are the same using colmena as I originally thought (before realizing that I mostly use nix-shell to only evaluate this package and test the results after) it could be relevant.

```
[ERROR]   stderr) error: hash mismatch in fixed-output derivation '/nix/store/dpb2xsnr0ya8qym3rxf6z5n2fykcygzl-android-tools-33.0.3.tar.xz.drv':
[ERROR]   stderr)          specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
[ERROR]   stderr)             got:    sha256-jOF02reB1d69Ke0PllciMfd3vuGbjvPBZ+M9PqdnC8U=
```

Reply:
Indeed. `fakeSha256` isn't outputting anything. Nix itself is. `fakeSha256` is just an unique value that no other derivation/store-path hashes to.

Older versions of Nix would indeed output sha256 hashes, but recent versions of Nix output SRI hashes, regardless of what was defined in the derivation. SRI hashes are a good thing (makes it clear in Nix's output that it is indeed a sha256 and not a sha512 hash), but the switch from sha256 to SRI is indeed confusing.

> So that's why I came up using this value/format. To mix it up like using sha256 = lib.fakeHash sound really wrong and I could see a benefits for newer users to technically disallowing it.

Yes, that totally makes sense how you came up with the value. I'll look into writing a RFC for this. Seems like an often-used feature that is very confusing for new users.

Reply:
Thanks for taking the time clarifying and following this up 👍🏻 :)